### PR TITLE
Add a confirmation modal window when deactivating Jetpack

### PR DESCRIPTION
**Pull request by leogermani**
Fixes #15858 

This PR is part of the efforts to better handle disconnection (see p1HpG7-8VJ-p2) and adds a confirmation modal window when the user tries to Deactivate Jetpack while other plugins are relying on the Jetpack connection.

### Background

Now that the Connection package is being used by other plugins, we want to avoid the situation where the user break other plugins when they deactivate Jetpack.

Since Jetpack clear the connection tokens when deactivated, this would break other active plugins relying on the connection. We want to warn users about this.

We've already added this warning to the "Manage site connection" dialog in the Jetpack Dashboard (see #15456) and, after some work with the design tem (p6TEKc-3FC-p2), we made this warning more prominent (#15818).

We agreed that the same warning should be shown at deactivation, and the original idea was to reuse the same modal. After working on #14965 we realized it would not be possible in the short run, and we have to rethink the whole disconnection flow. 

But since this warning is important, especially now with WCPayments being launched, we decided to go for a new modal and worked on this PR, building it from scratch and only displaying the list of active plugins (discarding all the other contents from the disconnection modal).

### Changes proposed in this Pull Request:

* In the plugins screen, if the user is connected and there is one or more other plugins using Jetpack connection, it:
 * Changes the label of the Deactivate link to "Disconnect and Deactivate"
 * When the user clicks it, it opens a modal window with a confirmation dialog displaying all other plugins that are using the connection and warning the user that they will lose connection if Jetpack is deactivated.

### Is this a new feature or does it add/remove features to an existing part of Jetpack?
* It adds a new confirmation step when the user is deactivating Jetpack (and has other plugins using connection)

### Challenges

The main challenge here is to make this modal look like the React `Modal` component used in the Jetpack Dashboard.

We are using Thickbox, and it will auto calculate the modal height when the window is opened and also at every screen resize, so it's hard to have full control of the height of the elements. 

Since the list of plugins will not grow in the near future, I favored a small window by default. It won't grow in height, so when the list of plugin grows, it will add a scrollbar (see screenshots).

#### Known issue

Thickbox changes the modal position at every window resize. I manage to override this in the first load, so I vertical centralized the modal, but if you resize the window, it will go back to the top. Also, I was not able to horizontal align it because Thickbox is doing this in a weird way, using a negative value for `margin-left` hard to calculate.

### Testing instructions:

* Set up a site using this branch and connect Jetpack
* Visit Plugins page on the Admin dashboard
* Check and see that the "Deactivate" link for the Jetpack plugin looks the same and nothing changed
* Create a file `wp-content/plugins/jetpack-connection-test-1.php` with following content:
```
<?php
/** Plugin Name: Jetpack Connection Test Plugin (1) */
add_action( 'plugins_loaded', function() {
	( new Automattic\Jetpack\Config() )->ensure( 'connection', array( 'slug' => 'jetpack-connection-test-1', 'name' => 'Jetpack Connection Test Plugin (1)' ) );
}, 1 );
```
* Activate this new plugin
* Now check that the label to deactivate Jetpack says "Disconnect and Deactivate"
* Click "Disconnect and Deactivate" and check that it opens a modal looking like the screenshots below
* Click Cancel and the close button at the top and confirm it closes the window
* Click the red "Disconnect and Deactivate" button and confirm it deactivates Jetpack

#### Layout tests

* Repeat the tests above resizing the window in different sizes
* Repeat the tests emulating mobile resolutions
* Repeat the tests in different browsers
* Tweak the HTML and make the list of plugins grow until it adds a vertical scrollbar to the modal.
 * Check again in all screen sizes with this extended list

#### Screenshots

Where you click to open the modal:
Captura de tela de 2020-05-19 19-11-17

Modal window opened:

Captura de tela de 2020-05-20 17-48-14

Big list with scroll:

Captura de tela de 2020-05-20 17-49-11

Captura de tela de 2020-05-20 17-49-24

Mobile:

photo_2020-05-21_12-06-51

### Proposed changelog entry for your changes:

* Adds a confirmation dialog upon deactivation when there are other plugins using connection

### CODE REVIEW

**Comment by folletto** [comment]
I had a quick chat with Leo and we clarified a few points as this was a bit too different from what we discussed in terms of flow.

Namely: the technical difficulties noted above means here we can't reuse the old modal at all, not just "reuse for this first step", but it won't load at all. So that means there's just a single step in the flow, and the second modal won't show.

This is something we'll have to review later, but it's the foundational assumption for this PR, so I take that as given.

—

I think this then is good, but needs some alignment tweaks.

fixes-align

1. Title margin on top, left, bottom should be all the same
2. The "X" should be similarly distanced equally from top, right, and bottom of the header
3. Left alignment of title, copy, bullet list, and "Are you sure" should be the same
4. Buttons at the bottom should be equally spaced top, right, and bottom, same measure as the header
5. The icon should be vertically centered with the text

**Comment by sergeymitr** [comment]
Hi @leogermani.
Good job on the modal, I like it a lot!

Just a few minor notes:
1. It looks a little bit strange that the modal is in the top of the page and not in the middle.
2. The white space in the modal is too large in my opinion, may be we should make the modal smaller?
deactivate-modal-small
3. An argument toward the (2) is that the modal doesn't fit into horizontally positioned small screens:
deactivate-modal-horizontal

**Comment by leogermani** [comment]
@sergeymitr 

Thanks for reviewing it.

The thing is that Thickbox plugin takes control over the modal height and position and it's pretty hard to override its default behavior in a nice way... I'll put more effort on it but any help is appreciated.

**Comment by leogermani** [comment]
@folletto I've worked on aligments and updated the screenshots

@sergeymitr I made the window smaller and centralized it. I took into consideration that the list of plugins will not be big in the near future, so it's ok to have a small fixed height.

There is one known issue though. Thickbox will send the window back to the top of the screen if you resize the window.

PR description updated

File: class.jetpack.php [hunk 1]
```diff
@@ -4161,6 +4163,58 @@ function plugin_action_links( $actions ) {
...
 	}
 
+	/**
+	 * Adds the deactivation warning modal if there are other active plugins using the connection
+	 *
+	 * @return void
+	 */
+	public function deactivate_dialog() {
+		global $pagenow;
+		if ( 'plugins.php' === $pagenow && self::is_active() ) {
```
Comment (jeherve) [inline class.jetpack.php:4173]: Instead of relying on the global, what would you think about using the info passed with the action?

```suggestion
	 *
	 * @param string $hook The current admin page.
	 *
	 * @return void
	 */
	public function deactivate_dialog( $hook ) {
		if (
			'plugins.php' === $hook
			&& self::is_active()
		) {
```

File: scss/jetpack-deactivate-dialog.scss [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+		max-height: 360px;
+	}
+
+	#TB_title {
+		font-size: 28px;
+		line-height: 24px;
+		font-weight: 600;
+		height: 24px;
+		padding: 24px;
+		color: #23282d;
```
Comment (jeherve) [inline scss/jetpack-deactivate-dialog.scss:16]: I would recommend avoiding hardcoding those colors, and instead rely on variables defined elsewhere when possible. I've added a few examples below, and I would encourage you to review the whole file for values you can use.

```suggestion
		color: $wpui-header-dark;
```

File: scss/jetpack-deactivate-dialog.scss [hunk 3]
```diff
@@ -0,0 +1,174 @@
+#TB_window.jetpack-disconnect-modal {
```
Comment (jeherve) [inline scss/jetpack-deactivate-dialog.scss:1]:
```suggestion
@import '../_inc/client/scss/variables/_colors.scss';

#TB_window.jetpack-disconnect-modal {
```

File: scss/jetpack-deactivate-dialog.scss [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+				li {
+
+					margin-bottom: 10px;
+
+					span {
+
+						vertical-align: middle;
+
+						&.dashicons {
+							color: red;
```
Comment (jeherve) [inline scss/jetpack-deactivate-dialog.scss:61]:
```suggestion
							color: $alert-hot-red;
```

File: scss/jetpack-deactivate-dialog.scss [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+			}
+
+
+		}
+	}
+
+	#TB_closeWindowButton {
+		left: auto;
+		right: 22px;
+		color: #a2a2a2;
```
Comment (jeherve) [inline scss/jetpack-deactivate-dialog.scss:77]:
```suggestion
		color: $wpui-dark-medium-gray;
```

Maybe? That's not the exact same color. I am not sure `#a2a2a2` in on the color palette of wp-admin?

File: views/admin/deactivation-dialog.php [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+?>
+<div id="jetpack_deactivation_dialog">
+
+	<div class="jetpack_deactivation_dialog_content">
+		<p>
+		<?php
+		echo esc_html(
+			sprintf(
+				// translators: %d is the number of additional plugins using the jetpack connection.
+				_n( // phpcs:ignore WordPress.WP.I18n.MismatchedPlaceholders
```
Comment (jeherve) [inline views/admin/deactivation-dialog.php:20]:
```suggestion
				/* translators: %d is the number of additional plugins using the jetpack connection. */
				_n(
```

**Comment by folletto** [comment]
From the screenshot design wise it's better. 
There's still some extra space on the mobile size (see that bottom margin is more than the other margins around the buttons?) but I think if it's not easy this can be merged for now. :)

**Comment by leogermani** [comment]
> From the screenshot design wise it's better.
> There's still some extra space on the mobile size (see that bottom margin is more than the other margins around the buttons?) but I think if it's not easy this can be merged for now. :)

Managed to fix it!

**Comment by leogermani** [comment]
@jeherve Thanks for the review! I didn't know about the variables. I managed to find them all and replaced it!

**Comment by kbrown9** [comment]
When I follow these steps, the modal content doesn't display:
1. Activate the test plugin.
2. Wait for the Plugins page to reload.
3. Click "Disconnect and Deactivate".

 If I navigate away from the Plugins page between steps 2 and 3, the modal content displays as expected. Have you seen this happen?

> Also, I was not able to horizontal align it because Thickbox is doing this in a weird way, using a negative value for margin-left hard to calculate.

I think we can change the `margin-left` value using something like:

  `$( '#TB_window' ).css( 'margin-left',  '?px');`

Or we can just remove the `margin-left` value with:

  `$( '#TB_window' ).css( 'margin-left', '' );`

I don't think the horizontal alignment is a blocker on this PR, but it would be nice to fix at some point.

**Comment by leogermani** [comment]
Thanks for reviewing it @kbrown9 !

> When I follow these steps, the modal content doesn't display:
> 
>     1. Activate the test plugin.
> 
>     2. Wait for the Plugins page to reload.
> 
>     3. Click "Disconnect and Deactivate".
> 
> 
> If I navigate away from the Plugins page between steps 2 and 3, the modal content displays as expected. Have you seen this happen?

I've seen this happen when there was a js error in the console about something completely unrelated. But hard to reproduce...

> 
> > Also, I was not able to horizontal align it because Thickbox is doing this in a weird way, using a negative value for margin-left hard to calculate.
> 
> I think we can change the `margin-left` value using something like:
> 
> `$( '#TB_window' ).css( 'margin-left', '?px');`
> 
> Or we can just remove the `margin-left` value with:
> 
> `$( '#TB_window' ).css( 'margin-left', '' );`
> 
> I don't think the horizontal alignment is a blocker on this PR, but it would be nice to fix at some point.

I can easily centralize the window removing the left margin and setting a `left` attribute, just the way I do with the `top`.

The problem with this is that Thickbox recalculates the modal position when you resize the window. When it does it, it doesn't touch the `left` attribute, only fills the left margin again. The result is the modal going out of the screen, because it sums up both `left` and `maring-left`.

So I tried to centralize using the `margin-left`. This is what thickbox does:

```JS
jQuery("#TB_window").css({marginLeft: '-' + parseInt((TB_WIDTH / 2),10) + 'px', width: TB_WIDTH + 'px'});
```

Basically adds a `margin-left` of half the modal width. It's not clear to me where is the starting point of the modal and why it used this approach instead of `left`... I'm not sure how to tweak the `margin-left` value in a consistent way to work in all screen sizes...

**Comment by kbrown9** [comment]
> Thanks for reviewing it @kbrown9 !
> 
> > When I follow these steps, the modal content doesn't display:
> > ```
> > 1. Activate the test plugin.
> > 
> > 2. Wait for the Plugins page to reload.
> > 
> > 3. Click "Disconnect and Deactivate".
> > ```
> > 
> > 
> > If I navigate away from the Plugins page between steps 2 and 3, the modal content displays as expected. Have you seen this happen?
> 
> I've seen this happen when there was a js error in the console about something completely unrelated. But hard to reproduce...
> 
This is strange; I see this happen every time I test on my ngrok test site, and there are no errors in the console.

Have you been able to test this on a JN site? I'm running this branch, and the "Deactivate" link isn't changing to "'Disconnect and Deactivate". The disconnect modal is working properly, so I know that the connected plugins list is being generated.

**Comment by leogermani** [comment]
> Have you been able to test this on a JN site? I'm running this branch, and the "Deactivate" link isn't changing to "'Disconnect and Deactivate". The disconnect modal is working properly, so I know that the connected plugins list is being generated.

It will only display the modal if there is another plugin using the connection.

I'll try to reproduce the content of the modal disappearing again.

I actually spent a couple of hours in the first day I worked on this because all of a sudden the content disappeared. And then, for no apparent reason, it appeared again... :/

**Comment by jeherve** [comment]
Looks like this is going to need a rebase.

**Comment by leogermani** [comment]
> Looks like this is going to need a rebase.

done

**Comment by jeherve** [comment]
> Have you been able to test this on a JN site? I'm running this branch, and the "Deactivate" link isn't changing to "'Disconnect and Deactivate". 

That's because on JN, the Jetpack plugin slug is `jetpack-dev`, not `jetpack`, so the link will not change.

**Comment by leogermani** [comment]
@jeherve @kbrown9 ,

Finally found out what was going on that the content would disappear sometimes. It was a problem building the Thickbox URL right after activating or deactivating plugins. When there are other parameters in the URL (such as `?plugin_status=all&paged=1&s` when you deactivate/activate a plugin), Thickbox will not parse its options correctly.

It's now fixed.
